File: backend/utils/redis_listener.py
import asyncio
import json
import os
from utils.websocket_manager import ws_manager
from utils.redis_memory_manager import memory_manager

import redis.asyncio as aioredis  # type: ignore

REDIS_URL = os.getenv("REDIS_URL", "redis://localhost:6379/0")

def clear_chat_memory(session_id: str, user_id: str | None = None) -> bool:
    """Clear Redis memory for a specific chat session"""
    try:
        import redis
        redis_client = redis.Redis.from_url(REDIS_URL)
        
        if user_id:
            # Clear specific user session
            pattern = f"*user_{user_id}_session_{session_id}*"
        else:
            # Clear all sessions with this ID
            pattern = f"*session_{session_id}*"
            
        keys = redis_client.keys(pattern)
        if keys:
            redis_client.delete(*keys)
            print(f"🧹 Cleared Redis memory: {len(keys)} keys for session {session_id}")
            return True
        return False
    except Exception as e:
        print(f"❌ Failed to clear Redis memory: {e}")
        return False

async def pubsub_listener():
    """Subscribe to mfu:chat:pubsub:* channels and forward messages to websockets."""
    redis = aioredis.from_url(REDIS_URL, decode_responses=True)
    pubsub = redis.pubsub()
    
    # Use the same pattern as memory manager
    pattern = f"{memory_manager.PUBSUB_PREFIX}:*"
    await pubsub.psubscribe(pattern)
    print(f"🔍 Redis listener subscribed to pattern: {pattern}")

    async for message in pubsub.listen():
        if message is None:
            continue
        if message["type"] not in ("pmessage", "message"):
            continue
        channel = message["channel"]  # pattern mfu:chat:pubsub:<session_id>
        data = message["data"]
        try:
            # extract session_id from mfu:chat:pubsub:<session_id>
            if isinstance(channel, bytes):
                channel = channel.decode()
            session_id = channel.split(":", 3)[3]  # Get the session_id part
            print(f"📨 Redis listener received message for session {session_id}: {data[:100]}...")
            await ws_manager.broadcast(session_id, data)
        except Exception as e:
            print(f"❌ Redis listener error: {e}") 

File: backend/src/utils/redisListener.ts
import { createClient } from 'redis';
import { config } from '../config/config';
import { wsManager } from './websocketManager';

export class RedisListener {
  private subscriber: any;
  private publisher: any;

  constructor() {
    this.subscriber = createClient({ url: config.redisUrl });
    this.publisher = createClient({ url: config.redisUrl });
    this.setupSubscriber();
  }

  private async setupSubscriber() {
    try {
      await this.subscriber.connect();
      await this.publisher.connect();
      console.log('✅ Redis listener connected');
      // Subscribe to chat channels
      await this.subscriber.pSubscribe('chat:*', async (message: string, channel: string) => {
        try {
          const sessionId = channel.replace('chat:', '');
          // broadcast message (string) to all ws clients in this session
          await wsManager.broadcastToSession(sessionId, message);
        } catch (error) {
          console.error('❌ Error broadcasting Redis message:', error);
        }
      });
    } catch (error) {
      console.error('❌ Redis listener connection error:', error);
    }
  }

  // Publish message to Redis
  async publish(channel: string, message: any) {
    try {
      await this.publisher.publish(channel, JSON.stringify(message));
    } catch (error) {
      console.error('❌ Redis publish error:', error);
    }
  }

  // Close connections
  async close() {
    try {
      await this.subscriber.disconnect();
      await this.publisher.disconnect();
      console.log('🔌 Redis listener disconnected');
    } catch (error) {
      console.error('❌ Error closing Redis listener:', error);
    }
  }
}

export const redisListener = new RedisListener(); 